File: backend/src/artifact_drafts.py
from __future__ import annotations

import json
from typing import Any


def unique_tags(tags: list[str | None]) -> list[str]:
    seen: set[str] = set()
    normalized: list[str] = []
    for tag in tags:
        if tag is None:
            continue
        candidate = tag.strip().lower()
        if not candidate or candidate in seen:
            continue
        seen.add(candidate)
        normalized.append(candidate)
    return normalized


def summarize(text: str, fallback: str) -> str:
    source = text.strip() or fallback.strip()
    return source[:180]


def stringify_output(output: Any) -> str:
    return json.dumps(output, ensure_ascii=True, indent=2)


def text_output(output: dict[str, Any]) -> str:
    content = output.get("content")
    return content.strip() if isinstance(content, str) else ""


def _default_artifact_type(studio: str) -> str:
    return {
        "writing": "document",
        "research": "report",
        "finance": "report",
        "data": "report",
        "image": "image",
    }.get(studio, "report")


def build_run_artifact_payload(
    *,
    run: dict[str, Any],
    title: str | None = None,
    artifact_type: str | None = None,
    summary: str | None = None,
    tags: list[str] | None = None,
    content_format: str | None = None,
) -> dict[str, Any] | None:
    output = run.get("output")
    if not isinstance(output, dict):
        return None

    studio = str(run.get("studio") or "report")
    run_title = str(run.get("title") or "Untitled Artifact")
    run_prompt = str(run.get("prompt") or run_title)
    chosen_title = (title or run_title).strip() or "Untitled Artifact"
    chosen_type = (artifact_type or _default_artifact_type(studio)).strip() or _default_artifact_type(studio)
    chosen_summary = summary or summarize(run_prompt, chosen_title)

    metadata = {
        "runId": run.get("id"),
        "model": run.get("model"),
        "status": run.get("status"),
        "prompt": run.get("prompt"),
        "options": run.get("options") or {},
        "output": output,
    }

    preview_image: str | None = None
    content_json: Any | None = None

    if studio == "image":
        urls = output.get("urls")
        if isinstance(urls, list) and urls:
            preview_image = str(urls[0])
        content = str(
            output.get("enhanced_prompt")
            or output.get("prompt")
            or run.get("prompt")
            or chosen_title
        )
        artifact_content_format = content_format or "plain"
        content_json = output
    else:
        content = text_output(output)
        if not content:
            content = stringify_output(output)
        artifact_content_format = content_format or ("markdown" if text_output(output) else "json")
        if artifact_content_format == "json":
            content_json = output

    combined_tags = unique_tags([studio, chosen_type, "saved-output", *(tags or [])])

    return {
        "title": chosen_title,
        "type": chosen_type,
        "studio": studio,
        "content": content,
        "summary": chosen_summary,
        "content_format": artifact_content_format,
        "metadata": metadata,
        "tags": combined_tags,
        "preview_image": preview_image,
        "content_json": content_json,
    }


def build_compare_artifact_payload(
    *,
    prompt: str,
    result: dict[str, Any],
    context_ids: list[str],
) -> dict[str, Any] | None:
    content = str(result.get("content") or "").strip()
    if not content:
        return None

    model = str(result.get("model") or "unknown-model")
    return {
        "title": f"Compare result: {model}",
        "type": "compare_result",
        "studio": "chat",
        "content": content,
        "summary": summarize(prompt, model),
        "content_format": "plain",
        "metadata": {
            "comparePrompt": prompt,
            "model": model,
            "latencyMs": result.get("latencyMs"),
            "status": result.get("status"),
            "contextIds": context_ids,
        },
        "tags": unique_tags(["chat", "compare", model]),
        "preview_image": None,
        "content_json": None,
    }


def build_image_artifact_payload(
    *,
    prompt: str,
    model: str,
    ratio: str,
    quality: str,
    url: str,
    storage_path: str | None = None,
) -> dict[str, Any] | None:
    if not url.strip():
        return None

    return {
        "title": prompt.strip()[:60] or "Generated image",
        "type": "image",
        "studio": "image",
        "content": prompt.strip() or "Generated image",
        "summary": f"Generated with {model}",
        "content_format": "plain",
        "metadata": {
            "model": model,
            "ratio": ratio,
            "quality": quality,
            "storage_path": storage_path or "",
        },
        "tags": unique_tags(["image", "generated"]),
        "preview_image": url,
        "content_json": None,
    }

File: frontend/src/lib/artifactDrafts.ts
import type { ChatMessage, CompareResult, CreateArtifactInput, DataAnalysisResult, RunRecord } from "./api";

function uniqueTags(tags: Array<string | null | undefined>): string[] {
  return [
    ...new Set(
      tags
        .map((tag) => tag?.trim().toLowerCase())
        .filter((tag): tag is string => Boolean(tag)),
    ),
  ];
}

function summarize(text: string, fallback: string) {
  const source = text.trim() || fallback.trim();
  return source.slice(0, 180);
}

function stringifyOutput(output: Record<string, unknown>) {
  return JSON.stringify(output, null, 2);
}

function textOutput(output: Record<string, unknown>) {
  return typeof output.content === "string" ? output.content.trim() : "";
}

function dataAnalysisToMarkdown(result: DataAnalysisResult) {
  const chart = result.chart_data;
  const firstDataset = chart.datasets[0];
  const chartRows = firstDataset
    ? chart.labels.map((label, index) => `| ${label} | ${firstDataset.data[index] ?? ""} |`).join("\n")
    : "";
  const resultRows = result.table.rows
    .map((row) => `| ${row.map((cell) => String(cell).replace(/\|/g, "\\|")).join(" | ")} |`)
    .join("\n");

  return [
    "## Insight",
    result.insight,
    "",
    "## Metrics",
    result.metrics?.length
      ? [
          "| Metric | Value | Note |",
          "| --- | --- | --- |",
          ...result.metrics.map((metric) => (
            `| ${String(metric.label).replace(/\|/g, "\\|")} | ${String(metric.value).replace(/\|/g, "\\|")} | ${String(metric.note ?? "").replace(/\|/g, "\\|")} |`
          )),
        ].join("\n")
      : "No decision metrics returned.",
    "",
    "## Risks",
    result.risks?.length
      ? [
          "| Risk | Severity | Evidence |",
          "| --- | --- | --- |",
          ...result.risks.map((risk) => (
            `| ${String(risk.risk).replace(/\|/g, "\\|")} | ${String(risk.severity).replace(/\|/g, "\\|")} | ${String(risk.evidence ?? "").replace(/\|/g, "\\|")} |`
          )),
        ].join("\n")
      : "No risks returned.",
    "",
    "## Recommendations",
    result.recommendations?.length
      ? result.recommendations.map((recommendation) => `- ${recommendation}`).join("\n")
      : "No recommendations returned.",
    "",
    "## Chart",
    `Type: ${result.chart_type}`,
    firstDataset
      ? [
          "",
          `Series: ${firstDataset.label}`,
          "",
          "| Label | Value |",
          "| --- | ---: |",
          chartRows,
        ].join("\n")
      : "No chart data returned.",
    "",
    "## Table",
    result.table.headers.length
      ? [
          `| ${result.table.headers.map((header) => String(header).replace(/\|/g, "\\|")).join(" | ")} |`,
          `| ${result.table.headers.map(() => "---").join(" | ")} |`,
          resultRows,
        ].join("\n")
      : "No table returned.",
  ].join("\n");
}

export function buildRunArtifactInput({
  studio,
  title,
  prompt,
  run,
  type = "report",
  tags = [],
}: {
  studio: string;
  title: string;
  prompt: string;
  run: RunRecord | null;
  type?: string;
  tags?: string[];
}): CreateArtifactInput | null {
  if (!run?.output) {
    return null;
  }

  const content = textOutput(run.output) || stringifyOutput(run.output);
  const contentFormat = textOutput(run.output) ? "markdown" : "json";

  return {
    title,
    type,
    studio,
    content,
    summary: summarize(prompt, title),
    content_format: contentFormat,
    metadata: {
      runId: run.id,
      model: run.model,
      status: run.status,
      prompt,
      options: run.options,
      output: run.output,
    },
    tags: uniqueTags([studio, type, "saved-output", ...tags]),
  };
}

export function buildDataArtifactInput({
  fileName,
  prompt,
  analysisResult,
  runId,
}: {
  fileName: string;
  prompt: string;
  analysisResult: DataAnalysisResult | null;
  runId?: string | null;
}): CreateArtifactInput | null {
  if (!analysisResult) return null;

  const title = `Data insight: ${fileName || "uploaded dataset"}`;
  return {
    title,
    type: "report",
    studio: "data",
    content: dataAnalysisToMarkdown(analysisResult),
    content_format: "markdown",
    summary: summarize(analysisResult.insight, title),
    content_json: analysisResult as unknown as Record<string, unknown>,
    source_run_id: runId ?? null,
    metadata: { prompt: prompt || title, fileName },
    tags: uniqueTags(["data", "insights", "chart", "table", "saved-output"]),
  };
}

export function buildChatMessageArtifactInput({
  threadTitle,
  message,
}: {
  threadTitle: string;
  message: ChatMessage;
}): CreateArtifactInput | null {
  if (message.role !== "assistant" || !message.content.trim()) {
    return null;
  }

  const titleSource = message.content.replace(/\s+/g, " ").trim();
  const title = `Chat output: ${titleSource.slice(0, 56) || threadTitle || "assistant response"}`;

  return {
    title,
    type: "chat_response",
    studio: "chat",
    content: message.content,
    content_format: "markdown",
    summary: summarize(message.content, threadTitle || title),
    metadata: {
      threadTitle,
      messageId: message.id,
      contextIds: Array.isArray(message.metadata?.contextIds) ? message.metadata.contextIds : [],
    },
    tags: uniqueTags(["chat", "assistant-output", "saved-output"]),
  };
}

export function buildDataChartArtifactInput({
  fileName,
  prompt,
  analysisResult,
  runId,
}: {
  fileName: string;
  prompt: string;
  analysisResult: DataAnalysisResult | null;
  runId?: string | null;
}): CreateArtifactInput | null {
  if (!analysisResult?.chart_data) return null;

  const chart = analysisResult.chart_data;
  const firstDataset = chart.datasets[0];
  const rows = firstDataset
    ? chart.labels.map((label, index) => `| ${label} | ${firstDataset.data[index] ?? ""} |`).join("\n")
    : "";
  const title = `Data chart: ${fileName || "uploaded dataset"}`;

  return {
    title,
    type: "chart",
    studio: "data",
    content: [
      `# ${title}`,
      "",
      `Type: ${analysisResult.chart_type}`,
      firstDataset ? `Series: ${firstDataset.label}` : "No series returned.",
      "",
      "| Label | Value |",
      "| --- | ---: |",
      rows,
    ].join("\n"),
    content_format: "markdown",
    summary: summarize(prompt, title),
    content_json: {
      chartType: analysisResult.chart_type,
      chartData: analysisResult.chart_data,
    },
    source_run_id: runId ?? null,
    metadata: { prompt: prompt || title, fileName },
    tags: uniqueTags(["data", "chart", analysisResult.chart_type, "saved-output"]),
  };
}

export function buildDataTableArtifactInput({
  fileName,
  prompt,
  analysisResult,
  runId,
}: {
  fileName: string;
  prompt: string;
  analysisResult: DataAnalysisResult | null;
  runId?: string | null;
}): CreateArtifactInput | null {
  if (!analysisResult?.table?.headers.length) return null;

  const table = analysisResult.table;
  const rows = table.rows
    .map((row) => `| ${row.map((cell) => String(cell).replace(/\|/g, "\\|")).join(" | ")} |`)
    .join("\n");
  const title = `Data table: ${fileName || "uploaded dataset"}`;

  return {
    title,
    type: "table",
    studio: "data",
    content: [
      `# ${title}`,
      "",
      `| ${table.headers.map((header) => String(header).replace(/\|/g, "\\|")).join(" | ")} |`,
      `| ${table.headers.map(() => "---").join(" | ")} |`,
      rows,
    ].join("\n"),
    content_format: "markdown",
    summary: summarize(prompt, title),
    content_json: table as unknown as Record<string, unknown>,
    source_run_id: runId ?? null,
    metadata: { prompt: prompt || title, fileName },
    tags: uniqueTags(["data", "table", "saved-output"]),
  };
}

export function buildWritingArtifactInput({
  title,
  content,
  summary,
  runId,
  contextIds = [],
}: {
  title: string;
  content: string;
  summary: string;
  runId?: string | null;
  contextIds?: string[];
}): CreateArtifactInput | null {
  if (!content.trim()) {
    return null;
  }

  return {
    title: title.trim() || "Untitled Document",
    type: "document",
    studio: "writing",
    content,
    content_format: "markdown",
    summary: summarize(summary, content),
    metadata: {
      runId: runId ?? null,
      contextIds,
    },
    tags: uniqueTags(["writing", "document", "saved-output"]),
  };
}

export function buildCompareArtifactInput({
  prompt,
  result,
  contextIds,
}: {
  prompt: string;
  result: CompareResult;
  contextIds: string[];
}): CreateArtifactInput | null {
  if (!result.content.trim()) {
    return null;
  }

  return {
    title: `Compare result: ${result.model}`,
    type: "compare_result",
    studio: "chat",
    content: result.content,
    content_format: "plain",
    summary: summarize(prompt, result.model),
    metadata: {
      comparePrompt: prompt,
      model: result.model,
      latencyMs: result.latencyMs,
      status: result.status,
      finishReason: result.finishReason ?? null,
      toolCalls: result.toolCalls ?? [],
      contextIds,
    },
    tags: uniqueTags(["chat", "compare", result.model]),
  };
}

export function buildImageArtifactInput({
  prompt,
  model,
  ratio,
  quality,
  url,
  storagePath,
  contextIds = [],
}: {
  prompt: string;
  model: string;
  ratio: string;
  quality: string;
  url: string;
  storagePath?: string;
  contextIds?: string[];
}): CreateArtifactInput | null {
  if (!url.trim()) {
    return null;
  }

  return {
    title: prompt.trim().slice(0, 60) || "Generated image",
    type: "image",
    studio: "image",
    content: prompt.trim(),
    summary: `Generated with ${model}`,
    content_format: "plain",
    preview_image: url,
    metadata: {
      model,
      ratio,
      quality,
      storage_path: storagePath ?? "",
      contextIds,
    },
    tags: uniqueTags(["image", "generated"]),
  };
}
